test/ARCHITECTURE.md

```diff
@@ -0,0 +1,267 @@
+# Architecture
+
+This document describes the architecture of Cockpit's browser integration
+tests. The tests should replicate how a normal user interacts with Cockpit this
+requires a test machine which can easily add have multiple disks or interfaces,
+reboot, interact with multiple machines on the same network and run potentially
+destructive test scenario's (e.g. installing/updating packages, formatting
+disks).
+
+For these reasons, Cockpit tests run inside a virtual machine (VM). The virtual
+machine uses Cockpit specific virtual machine images maintained and created in
+the [bots](https://github.com/cockpit-project/bots) repository. The images are
+usually based on a distribution's cloud image customized with:
+
+* A well-known password for the admin/root user
+* Test SSH keys for access
+* Test packages required to test Cockpit
+* A build chroot with Cockpit's build dependencies to build the to be tested
+  Cockpit source inside the virtual machine offline. This allows a developer on
+  Fedora to easily prepare a Debian test image without having to install Debian
+  build tools.
+* Disabling system services which interfere with testing
+
+To replicate a user, Cockpit is tested in a browser controlled using the
+[Chrome DevTools Protocol](https://chromedevtools.github.io/devtools-protocol/)
+(CDP) which is supported by Firefox and Chromium based browsers.
+
+The Python test framework in `test/common` is responsible for setting up the
+test environment, running tests and reporting of the test output.
+
+Diagram of the interaction of Browser/Machine/CDP/Test Framework.
+```mermaid
+graph TD;
+    id[Test Framework] <-->|CDP| Browser;
+    id[Test Framework] <-->|SSH| id1[Virtual Machine];
+```
+
+## Integration Tests
+
+Cockpit's tests can be run via three different entry points:
+
+* `test/verify/check-$page` - run a single or multiple unit test(s)
+* `test/common/run-tests` - run tests through our test scheduler (retries, tracks naughties)
+* `test/run` - run tests in Continuous Integration (CI)
+
+We will start with how a single integration test is run and then explore the test
+scheduler and CI setup.
+
+### Test runtime
+
+The base of a Cockpit integration test looks as following:
+
+```python
+class TestApps(testlib.MachineCase):
+    def testBasic(self):
+        self.machine.execute("rm /usr/share/cockpit/apps")
+        self.browser.login_and_go("/apps")
+        self.browser.wait_not_visible("#apps")
```

Comment:
The login will fail because /apps doesn't exist. Also, a wait_not_visible() is a no-op or race condition. I. e. this example is very confusing and broken. This needs to stay on the shell, check that some other menu item is visible, and *then* check that Apps doesn't exist.